Improve fqdn events logging management

pippolo84 opened the pull request:
Currently, the `Log` method of fqdn related `LogRecord` is holding the lock for the entire record operation.
The record operation can be broken up in the following parts:

- add metadata to the record
- check if the notifier is set
- call `NewProxyLogRecord` on the notifier if not nil

The lock is needed both to copy metadata and to get the currently set notifier.
Instead, the call to `NewProxyLogRecord` can be executed without holding the lock, given that each concrete type that satisfies the `LogRecordNotifier` interface supports concurrent notifications.

Currently there is only one concrete implementation of that interface, the agent monitor. The agent monitor supports two kinds of clients: listeners and consumers.
Since listeners expect the notifications to be first JSON marshaled and then Gob encoded, the log record operation is not negligible both in terms of time taken and computational resources.

So, in case of:

- a very high rate of fqdn related events
- low computational resources available for the pod

the queue of goroutines waiting on that lock may grow significantly. In that scenario, the total memory used because of all the enqueued goroutines may grow unacceptably.

As an example, this can be observed when cilium agent pods are assigned limited computational resources and the rate of DNS requests is very high but consistently failing for a prolonged period of time.

First, this PR avoid the encoding if there are no active listeners.
Then, since the agent monitor already supports concurrent events handling, the commit shortens the critical section, allowing for concurrent execution of the actual log entries recording.
Doing so, the number of goroutines waiting for the `LogRecord` lock is lowered in high load - low resources scenarios, both with and without active listeners.

pippolo84 committed:
monitor: Do not marshal events without active listeners

Agent monitor supports both consumers and listeners. Unlike consumers,
a listener expects an event to be first JSON encoded, then binary
encoded using Gob.

Since the whole encoding operation may be expensive, we avoid that if
there are no active listeners at the moment.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
logger: Avoid holding the lock for the entire log operation

Currently, the Log() method is holding the LogRecord lock for the entire
record operation. The record operation can be broken up in the following
parts:

- add metadata to the record
- check if the notifier is set
- call NewProxyLogRecord on the notifier if not nil

The lock is needed both to copy metadata and to get the currently set
notifier.  Instead, the call to NewProxyLogRecord can be executed
without holding the lock if the concrete type that satisfies the
LogRecordNotifier interface supports concurrent notifications.

Currently there is only one concrete implementation of that interface:
the agent monitor. The agent monitor supports two kinds of clients:
listeners and consumers. Since listeners expect the notifications to be
first JSON marshaled and then Gob encoded, the log record operation is
not negligible both in terms of time taken and computational resources.

So, in case of:

- a very high rate of fqdn related events
- low computational resources available for the pod

the queue of goroutines waiting on that lock may grow significantly. In
that scenario, the total memory used because of all the enqueued
goroutines may grow unacceptably.

Since the agent monitor already supports concurrent events handling, the
commit shortens the critical section, allowing for concurrent execution
of the actual log entries recording. Doing so, the number of goroutines
waiting for the LogRecord lock is lowered in high load/low resources
scenarios.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

christarazi commented on pkg/monitor/agent/agent.go:
```diff
@@ -128,6 +128,11 @@ func (a *Agent) SendEvent(typ int, event interface{}) error {
+	if !a.hasListeners() {
+		return nil
+	}
+
```
I assume that we do the check for listeners after `notifyAgentEvent` because we want to ensure we still run some logic? Would be good to clarify why in the code, so that nobody attempts to optimize this with the same thought that it can be moved before `notifyAgentEvent`.

gandro replied:
Yeah.. to elaborate on the current terminology: Consumers (invoked by `notifyAgentEvent`) consume in-memory data structures, while Listeners (invoked by `sendToListeners` below) expected the data to already be marshaled. The two queues are separate, so we still want to invoke any Consumers even if there are no listeners.

Fabio and I discussed a bit if/how we could move the Listeners over to the Consumer interface and it should be possible, the tricky part is that we want to avoid marshaling the data for every Listener individually (and only marshal once for all listeners, as the current code does).

pippolo84 replied:
@christarazi probably it'd be good to expand the comment here a bit, to point out that we don't want to waste cycles marshalling if there are no active listeners. I'll add some lines to clarify the differences between consumers and listeners (and why the latter are much more expensive).

christarazi commented on pkg/proxy/logger/logger.go:
```diff
@@ -232,18 +232,19 @@ func (lr *LogRecord) Log() {
-	// NewProxyLogRecord is called for each new log record
+	// NewProxyLogRecord is called for each new log record.
+	// It may be called concurrently by multiple logger goroutines.
 	NewProxyLogRecord(l *LogRecord) error
```
I think you can go a bit further here and say that any object which wishes to implement the `LogRecordNotifier` interface, must support concurrent calls to the interface methods, and not rely on locks.

christarazi reviewed (commented):
Great find! It would be nice if we were able to create a benchmark test for this code.

gandro reviewed (approved):
Thanks!

pippolo84 committed:
logger: Add benchmarks for events log notifications

The benchmarks evaluate the impact of the previous two commits in
improving the events log notification performance.

Specifically, the benchmarks mimics what is happening in case of a high
rate of DNS related notifications and consequent logging of those
events, as done in the `notifyOnDNSMsg` callback.

To better understand the scalability improvements, the benchmarks run
with various numbers of concurrent notifications: from a single one up
to 1000.

The first benchmark, called LogNotifierWithNoListeners, shows the
improvement that comes from commit `monitor: Do not marshal events
without active listeners`, that avoids encoding notifications altogether
if no listeners are currently active.  As expected, the improvement is
remarkable, since no (useless) encoding is taking place at all.

The second one, called LogNotifierWithListeners, shows the improvement
that comes from commit `logger: Avoid holding the lock for the entire
log operation`, where we release the lock earlier to enable concurrent
log notifications encoding.  As expected, the performance improvement is
more evident as the load increases, thanks to the higher level of
concurrency.

Below the results of running these benchmarks without the two commits
(shown as `old`) and with them (shown as `new`), compared with
benchstat.

$ benchstat old.txt new.txt
name                                          old time/op    new time/op    delta
LogNotifierWithNoListeners/OneRecord-8          27.8µs ± 1%     2.6µs ± 3%  -90.49%  (p=0.008 n=5+5)
LogNotifierWithNoListeners/TenRecords-8          284µs ± 1%      34µs ±12%  -88.03%  (p=0.008 n=5+5)
LogNotifierWithNoListeners/HundredRecords-8     2.80ms ± 1%    0.21ms ± 4%  -92.65%  (p=0.008 n=5+5)
LogNotifierWithNoListeners/ThousandRecords-8    28.2ms ± 1%     1.8ms ± 1%  -93.63%  (p=0.016 n=5+4)
LogNotifierWithListeners/OneRecord-8            33.5µs ± 6%    48.6µs ± 2%  +45.13%  (p=0.008 n=5+5)
LogNotifierWithListeners/TenRecords-8            352µs ± 1%     330µs ± 1%   -6.30%  (p=0.016 n=5+4)
LogNotifierWithListeners/HundredRecords-8       3.47ms ± 2%    2.92ms ± 3%  -16.06%  (p=0.008 n=5+5)
LogNotifierWithListeners/ThousandRecords-8      35.3ms ± 1%    31.5ms ± 2%  -10.83%  (p=0.008 n=5+5)

name                                          old alloc/op   new alloc/op   delta
LogNotifierWithNoListeners/OneRecord-8          10.2kB ± 0%     0.9kB ± 0%     ~     (p=0.079 n=4+5)
LogNotifierWithNoListeners/TenRecords-8          102kB ± 0%       9kB ± 0%  -91.19%  (p=0.008 n=5+5)
LogNotifierWithNoListeners/HundredRecords-8     1.02MB ± 0%    0.09MB ± 0%  -91.20%  (p=0.008 n=5+5)
LogNotifierWithNoListeners/ThousandRecords-8    10.2MB ± 0%     0.9MB ± 0%  -91.25%  (p=0.008 n=5+5)
LogNotifierWithListeners/OneRecord-8            14.4kB ± 0%    14.4kB ± 0%   +0.05%  (p=0.024 n=5+5)
LogNotifierWithListeners/TenRecords-8            143kB ± 0%     123kB ± 0%  -13.92%  (p=0.008 n=5+5)
LogNotifierWithListeners/HundredRecords-8       1.43MB ± 0%    1.17MB ± 1%  -18.51%  (p=0.008 n=5+5)
LogNotifierWithListeners/ThousandRecords-8      14.3MB ± 0%    11.1MB ± 1%  -22.37%  (p=0.008 n=5+5)

name                                          old allocs/op  new allocs/op  delta
LogNotifierWithNoListeners/OneRecord-8             169 ± 0%         9 ± 0%  -94.67%  (p=0.008 n=5+5)
LogNotifierWithNoListeners/TenRecords-8          1.69k ± 0%     0.08k ± 0%  -95.20%  (p=0.008 n=5+5)
LogNotifierWithNoListeners/HundredRecords-8      16.9k ± 0%      0.8k ± 0%  -95.26%  (p=0.008 n=5+5)
LogNotifierWithNoListeners/ThousandRecords-8      169k ± 0%        8k ± 0%  -95.28%  (p=0.008 n=5+5)
LogNotifierWithListeners/OneRecord-8               196 ± 0%       196 ± 0%     ~     (all equal)
LogNotifierWithListeners/TenRecords-8            1.95k ± 0%     1.83k ± 0%   -6.53%  (p=0.016 n=4+5)
LogNotifierWithListeners/HundredRecords-8        19.5k ± 0%     17.8k ± 0%   -8.68%  (p=0.008 n=5+5)
LogNotifierWithListeners/ThousandRecords-8        195k ± 0%      175k ± 0%  -10.57%  (p=0.008 n=5+5)

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

christarazi reviewed (approved):
Thanks! Amazing results! :rocket:
